### DISCUSSION TITLE
Feature request for IoSender (sizing limits)

### DISCUSSION THREAD

**Discussion by Klippert** [terjeio/grblHAL#218] [Ideas]
I was wondering if it would be possible to do a "sizing". Currently in 3d view i can see the program limits, and I can manually jog to the X and Y limit, at a safe Z height, to see that i'm clear of clamps and such before starting a job. Would it be possible to make a small feature that does a slow-ish walk around the edge of the program loaded (without changing Z-height)?

**Comment by terjeio** [comment]
It is not possible to add this to grblHAL. However, it can be added to ioSender. What feedrate do you define as "slow-ish"?

**Reply by Klippert** [reply]
3-500mm/min or something like that. so one can actually see that one doesn't crash into anything :-)

**Comment by phomann** [comment]
I presume this feature is for checking the program without actually cutting?

If it was to be added to ioSender, maybe add a Z-limit function enable button and a Z-limit value field. That way when the function is enabled ioSender could limit the Z value sent to the value in Z-Limit field?

As to the Z feedrate, the feedrate override function could be used to slow down the movement.
This would allow for the original feature request but also other aspects as well.

Cheers,

Peter

**Reply by Klippert** [reply]
Yes, it's just to see visually on the machine that your toolpath is clear of clamps etc. 

Imho Z can just stay exactly where you are - i.e. you jog to the z-height you want, and hit the button, and it does a little walk around the edge (X and Y), and one can quickly see if the path is clear or something is uncomfortably close. just ran into another clamp today while doing a job, because i mistaenly thought it was far enough away. :-/ (endmill survived luckily).

Also a rookie question - what's the easiest Gcode to go to a coordinate with full feedrate? - ie. move fast to x + 100mm ? (machine coordinate)

**Reply by terjeio** [reply]
> Also a rookie question - what's the easiest Gcode to go to a coordinate with full feedrate? - ie. move fast to x + 100mm ? 

G91G0X100
G90

**Reply by Klippert** [reply]
thanks :-)

**Comment by phil-barrett** [comment]
If the intent is to make sure that your gcode and workpiece are co-incident, then Program Limits in the 3D view is pretty helpful.

**Reply by Klippert** [reply]
Yep, but to really see if i'm clear or not, i usually need to jog to the positions - that's where it would be handy to have the machine do a little walk around for me :-)

**Comment by terjeio** [comment]
Added as a flyout to the latest [edge version](http://io-engineering.com/downloads). Note that this has changes to  _Tool length offset_ probing that I hope is ok, please be careful when using this and report any issues!

**Comment by Klippert** [comment]
Thank you again Terjeio!
It didn't exactly work as i thought it would, but ther might be something i'm doing wrong.
I home the machine
i load up a program 
i set a X and Y position for my 0,0 of the program (through the DRO). 
![image](https://user-images.githubusercontent.com/53910032/106353225-3f125480-62e9-11eb-8fab-7ef2bcad8f96.png)

then i make sure my bit is lifted well above anything (in this case i just left it at it's home position).
this case: 
![image](https://user-images.githubusercontent.com/53910032/106353247-6cf79900-62e9-11eb-9af4-3dc30e77ece2.png)

open the outline:
![image](https://user-images.githubusercontent.com/53910032/106353257-7d0f7880-62e9-11eb-9dd7-5b5247784701.png)

I  hit go, and get an alarm 2. so i somehow exceeded my machines workable area.
![image](https://user-images.githubusercontent.com/53910032/106353325-f7d89380-62e9-11eb-86f3-6b9f95597121.png)

I'm not sure what gives.

**Reply by Klippert** [reply]
when i look at the program there is something that doesn't add up... look at this:
![image](https://user-images.githubusercontent.com/53910032/106353818-96b2bf00-62ed-11eb-95cc-eb418fc23ae0.png)
We're looking at a simple two pass facing operation, of a part that's not terribly large. (about 70x110mm)
I don't understand where it gets the -84 on the x-axis from and the -88 on the y axis. 

It does go a bit across 0 on the x axis, but Y should be all positives. 
Any suggestions?

below is the NC, it's not horribly long.
(Motormount-facing)
(T35  D=8 CR=0 - ZMIN=8 - flat end mill)
G90 G94
G17
G21
G28 G91 Z0
G90

(Face-Rough)
T35 M6
S12000 M3
G54
G0 X68.8 Y-0.484
Z25
Z15
G1 Z9 F1800
G18 G3 X68 Z8.2 I-0.8 K0
G1 X-14
G17 G2 X-14 Y6.957 I0 J3.72
G1 X68
G3 X68 Y14.398 I0 J3.72
G1 X-14
G2 X-14 Y21.838 I0 J3.72
G1 X68
G3 X68 Y29.278 I0 J3.72
G1 X-14
G2 X-14 Y36.719 I0 J3.72
G1 X68
G3 X68 Y44.16 I0 J3.72
G1 X-14
G2 X-14 Y51.6 I0 J3.72
G1 X68
G3 X68 Y59.041 I0 J3.72
G1 X-14
G2 X-14 Y66.481 I0 J3.72
G1 X68
G3 X68 Y73.922 I0 J3.72
G1 X-14
G2 X-14 Y81.362 I0 J3.72
G1 X68
G3 X68 Y88.802 I0 J3.72
G1 X-14
G2 X-14 Y96.243 I0 J3.72
G1 X68
G3 X68 Y103.684 I0 J3.72
G1 X-14
G2 X-14 Y111.124 I0 J3.72
G1 X68
G18 G2 X68.8 Z9 I0 K0.8
G0 Z25
G17

(Face1 - smoothing)
G0 X68.8 Y-3.989
Z25
Z15
G1 Z8.8 F1800
G18 G3 X68 Z8 I-0.8 K0
G1 X-14
G17 G2 X-14 Y-0.019 I0 J1.985
G1 X68
G3 X68 Y3.952 I0 J1.985
G1 X-14
G2 X-14 Y7.923 I0 J1.985
G1 X68
G3 X68 Y11.893 I0 J1.985
G1 X-14
G2 X-14 Y15.864 I0 J1.985
G1 X68
G3 X68 Y19.835 I0 J1.985
G1 X-14
G2 X-14 Y23.805 I0 J1.985
G1 X68
G3 X68 Y27.776 I0 J1.985
G1 X-14
G2 X-14 Y31.747 I0 J1.985
G1 X68
G3 X68 Y35.717 I0 J1.985
G1 X-14
G2 X-14 Y39.688 I0 J1.985
G1 X68
G3 X68 Y43.659 I0 J1.985
G1 X-14
G2 X-14 Y47.629 I0 J1.985
G1 X68
G3 X68 Y51.6 I0 J1.985
G1 X-14
G2 X-14 Y55.571 I0 J1.985
G1 X68
G3 X68 Y59.541 I0 J1.985
G1 X-14
G2 X-14 Y63.512 I0 J1.985
G1 X68
G3 X68 Y67.483 I0 J1.985
G1 X-14
G2 X-14 Y71.453 I0 J1.985
G1 X68
G3 X68 Y75.424 I0 J1.985
G1 X-14
G2 X-14 Y79.395 I0 J1.985
G1 X68
G3 X68 Y83.365 I0 J1.985
G1 X-14
G2 X-14 Y87.336 I0 J1.985
G1 X68
G3 X68 Y91.307 I0 J1.985
G1 X-14
G2 X-14 Y95.277 I0 J1.985
G1 X68
G3 X68 Y99.248 I0 J1.985
G1 X-14
G2 X-14 Y103.219 I0 J1.985
G1 X68
G3 X68 Y107.189 I0 J1.985
G1 X-14
G2 X-14 Y111.16 I0 J1.985
G1 X68
G18 G2 X68.8 Z8.8 I0 K0.8
G0 Z25
G17
G28 G91 Z0
G90
G28 G91 X0 Y0
G90
M5
M30

**Reply by terjeio** [reply]
It seems like a bug in my gcode parser - I guess due to plane changes (G17/G18). Will look into this later - I am off for some shopping.

**Reply by terjeio** [reply]
Looks like it is due to incorrect handling of G28 moves, new edge version uploaded that should work.

**Reply by Klippert** [reply]
I'll check that out shortly. I've just redone my fixture plate with a million holes with threading it took quite a while. But everything was very stable throughout this 2.5h job. I'm very pleased with that.

**Reply by Klippert** [reply]
and it works like a charm! Thanks! that feature is going to save me quite a few clamps and endmills :-)
